### DISCUSSION TITLE
stories_things/cpp-things/vptr_vtable/

### DISCUSSION THREAD

**Discussion by giscus** [Light-City/light-comments#10] [Announcements]
# stories_things/cpp-things/vptr-vtable/

Stories About C Plus Plus

https://light-city.github.io/stories_things/cpp-things/vptr-vtable/

**Comment by cxz0218** [comment]
作者大大，第一段代码的运行结果第三行应该是：基类对象调用基类实例~

**Reply by Light-City** [reply]
这里应该有问题 我下来抽时间改一下

**Comment by cxz0218** [comment]
感谢~，大佬写的真好




日光沉寂的半海
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Light-City/light-comments"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年3月26日(星期二) 下午3:09
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Light-City/light-comments] stories_things/basic_content/vptr_vtable/ (Discussion #10)





 
这里应该有问题 我下来抽时间改一下

**Reply by Light-City** [reply]
不客气 感谢支持～

**Comment by xllzs** [comment]
void* func_addr = (void *)*((unsigned long *)vptr_addr+offset);
offset = 0 的时候，请问vptr_addr不应该等于func_addr吗

vptr_addr:0x401130
func_addr:0x400ea8

**Reply by JasonPhone** [reply]
`(void *)*((unsigned long *)vptr_addr+offset)`，中间做了一次解引用。

**Reply by ThinkWithZ** [reply]
*(p + offset) 就是 p[0]，所以就是取虚表首个元素的地址

**Comment by ThinkWithZ** [comment]
基类析构函数应为虚函数virtual ~Base() {}
